### PR TITLE
document eachrow and eachcol

### PR DESCRIPTION
**Pull request by bkamins**
Improves documentation of `eachrow` and `eachcol`.
Additionally specifies `eltype` of `DFColumnIterator` to be `Tuple{Symbol, AbstractVector}`.

### CODE REVIEW

**Comment by pdeffebach** [comment]
One thing that would be nice is a mention of how `map` works. `map(f, eachcol(df)` returns a DataFrame, and `f` acts on the `col` rather than the tuple of `(name, col)`, which is slightly unintuitive, though superior behavior.

**Comment by bkamins** [comment]
@pdeffebach added

File: docs/src/lib/types.md [hunk 1]
```diff
@@ -33,6 +33,10 @@ and reflects changes done to the parent after the creation of the view.
 Typically objects of the `DataFrameRow` type are encountered when returned by the `eachrow` function.
 In the future accessing a single row of a data frame via `getindex` or `view` will return a `DataFrameRow`.
 
+Additionally `eachrow` and `eachcol` functions return values of `DFRowIterator` and `DFColumnIterator` types respectively.
```
Comment (nalimilan) [inline docs/src/lib/types.md]:
```suggestion
Additionally the `eachrow` and `eachcol` functions return values of the `DFRowIterator` and `DFColumnIterator` types respectively.
```

File: src/abstractdataframe/iteration.jl [hunk 2]
```diff
@@ -7,14 +7,23 @@
 # TODO: Reconsider/redesign eachrow -- ~100% overhead
 
 # Iteration by rows
+"""
+    DFRowIterator{<:AbstractDataFrame}
+
+Iterator over rows of an `AbstractDataFrame`.
+Each returned value is represented as a `DataFrameRow`.
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]: I'd say "with each row represented as a `DataFrameRow`" (as below), in a single sentence.

File: src/abstractdataframe/iteration.jl [hunk 3]
```diff
@@ -7,14 +7,23 @@
...
+
+A value of this type is returned by `eachrow` function.
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]:
```suggestion
A value of this type is returned by the [`eachrow`](@link) function.
```

File: src/abstractdataframe/iteration.jl [hunk 4]
```diff
@@ -7,14 +7,23 @@
...
+"""
 struct DFRowIterator{T <: AbstractDataFrame}
     df::T
 end
+
 """
-    eachrow(df) => DataFrames.DFRowIterator
+    eachrow(::AbstractDataFrame)
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]:
```suggestion
    eachrow(df::AbstractDataFrame)
```

File: src/abstractdataframe/iteration.jl [hunk 5]
```diff
@@ -29,16 +38,42 @@ Base.getindex(itr::DFRowIterator, i) = DataFrameRow(itr.df, i)
 Base.map(f::Function, dfri::DFRowIterator) = [f(row) for row in dfri]
 
 # Iteration by columns
+"""
+    DFColumnIterator{<:AbstractDataFrame}
+
+Iterator over columns of an `AbstractDataFrame`.
+Each returned value is a tuple consisting of column name and column data.
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]:
```suggestion
Each returned value is a tuple consisting of column name and column vector.
```

File: src/abstractdataframe/iteration.jl [hunk 6]
```diff
@@ -29,16 +38,42 @@ Base.getindex(itr::DFRowIterator, i) = DataFrameRow(itr.df, i)
...
+
+A value of this type is returned by `eachcol` function.
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]:
```suggestion
A value of this type is returned by the [`eachcol`](@link) function.
```

File: src/abstractdataframe/iteration.jl [hunk 7]
```diff
@@ -29,16 +38,42 @@ Base.getindex(itr::DFRowIterator, i) = DataFrameRow(itr.df, i)
...
+"""
 struct DFColumnIterator{T <: AbstractDataFrame}
     df::T
 end
+
+"""
+    eachcol(::AbstractDataFrame)
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]:
```suggestion
    eachcol(df::AbstractDataFrame)
```

File: src/abstractdataframe/iteration.jl [hunk 8]
```diff
@@ -29,16 +38,42 @@ Base.getindex(itr::DFRowIterator, i) = DataFrameRow(itr.df, i)
...
+
+Return `DFColumnIterator` that iterates an `AbstractDataFrame` column by column.
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]:
```suggestion
Return a `DFColumnIterator` that iterates an `AbstractDataFrame` column by column.
```

File: src/abstractdataframe/iteration.jl [hunk 9]
```diff
@@ -29,16 +38,42 @@ Base.getindex(itr::DFRowIterator, i) = DataFrameRow(itr.df, i)
...
+Iteration returns a tuple consisting of column name and column data.
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]:
```suggestion
Iteration returns a tuple consisting of column name and column vector.
```

File: src/abstractdataframe/iteration.jl [hunk 10]
```diff
@@ -29,16 +38,42 @@ Base.getindex(itr::DFRowIterator, i) = DataFrameRow(itr.df, i)
...
+
+`DFColumnIterator` has a custom implementation of the `map` function which
+returns a `DataFrame` and assumes that a function argument passed do
+the `map` function accepts only column data.
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]:
```suggestion
the `map` function accepts takes only a column vector.
```

File: src/abstractdataframe/iteration.jl [hunk 11]
```diff
@@ -29,16 +38,42 @@ Base.getindex(itr::DFRowIterator, i) = DataFrameRow(itr.df, i)
...
+
+**Examples**
+
+```julia
+df = DataFrame(x=1:4, y=11:14)
+map(sum, eachcol(df)) == DataFrame(x=10, y=50)
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]: Better use a `jldoctest` and copy the REPL output.

**Comment by nalimilan** [comment]
> One thing that would be nice is a mention of how `map` works. `map(f, eachcol(df)` returns a DataFrame, and `f` acts on the `col` rather than the tuple of `(name, col)`, which is slightly unintuitive, though superior behavior.

That's why I think we should consider exporting `columns`, and deprecate the behavior of `map` on `eachcol`, which is inconsistent.

File: src/abstractdataframe/iteration.jl (hunk 2 above)
Comment (bkamins) [reply src/abstractdataframe/iteration.jl]: fixed

File: src/abstractdataframe/iteration.jl (hunk 11 above)
Comment (bkamins) [reply src/abstractdataframe/iteration.jl]: fixed

File: src/abstractdataframe/iteration.jl [hunk 12]
```diff
@@ -7,14 +7,23 @@
...
     df::T
 end
+
 """
-    eachrow(df) => DataFrames.DFRowIterator
+    eachrow(df::AbstractDataFrame)
 
-Iterate a DataFrame row by row, with each row represented as a `DataFrameRow`,
-which is a view that acts like a one-row DataFrame.
+Return `DFRowIterator` that iterates an `AbstractDataFrame` row by row,
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]:
```suggestion
Return a `DFRowIterator` that iterates an `AbstractDataFrame` row by row,
```

File: src/abstractdataframe/iteration.jl [hunk 13]
```diff
@@ -28,16 +37,52 @@ Base.length(itr::DFRowIterator) = size(itr.df, 1)
...
+4×2 DataFrame
+│ Row │ x     │ y     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 11    │
+│ 2   │ 2     │ 12    │
+│ 3   │ 3     │ 13    │
+│ 4   │ 4     │ 14    │
+
+julia> map(sum, eachcol(df)) == DataFrame(x=10, y=50)
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl]: I mean that you can just print the output (it's weird to check for equality).

Comment (bkamins) [reply src/abstractdataframe/iteration.jl]: ok

**Pull request merged**